File: syseval-report.pl
#!/usr/bin/perl

#
# What's this?
# ------------
#
# Given a CSV-format manual evaluation file created by syseval-combine.pl,
# this script restores the orders of the systems and calculates an average
# score for each system.
#
# Usage and Data Format
# ---------------------
#
# syseval-report.pl finished-eval.csv output.ids
#
# Where finished-eval is a completely filled in manual evaluation CSV file
# and output.ids is the ID file output by syseval-combine.pl

use strict;
use warnings;
use utf8;
use Getopt::Long;
use List::Util qw(sum min max shuffle);
binmode STDIN, ":utf8";
binmode STDOUT, ":utf8";
binmode STDERR, ":utf8";

my $LINES = -1;
my $COMPS = "0-1";
GetOptions(
    "lines=i" => \$LINES,
    "comps=s" => \$COMPS,
);

if(@ARGV != 2) {
    print STDERR "Usage: $0 TSV IDS\n";
    exit 1;
}
open FILE0, "<:utf8", $ARGV[0] or die "Couldn't open $ARGV[0]\n";
open FILE1, "<:utf8", $ARGV[1] or die "Couldn't open $ARGV[1]\n";

my %lettermap = (
    "S" => 5,
    "AA" => 5,
    "A" => 4,
    "B" => 3,
    "C" => 2,
    "D" => 1,
    "F" => 1
);

my ($stsv, $sids, $lines, @scores, @tsvs, @vals, @refs, $header);
while(defined($stsv = <FILE0>) and defined($sids = <FILE1>)) {
    if((not $header) and ($stsv =~ /^(Source|Reference)\t/)) {
        $header = 1; $stsv = <FILE0>;
    }
    next if($stsv =~ /\tX\t/ or $stsv =~ /\tX$/);
    ++$lines;
    last if ($LINES != -1) and ($lines > $LINES);
    chomp $stsv; chomp $sids;
    push @tsvs, $stsv;
    $stsv =~ s/[\t ]+$//g;
    my @atsv = split(/\t/, $stsv);
    my @aids = split(/\t/, $sids);
    $refs[$lines-1] = shift(@atsv);
    $refs[$lines-1] .= "\t".shift(@atsv) if(@atsv % 2 == 1);
    if((max(@aids)+1)*2 != @atsv) { die "MISMATCHED LINES: ". (max(@aids)+1)*2 ." != ".scalar(@atsv)."\n$stsv\n$sids\n"; }
    foreach my $i (0 .. $#aids) {
        $scores[$i] = [] if not $scores[$i];
        my $pos = $aids[$i]*2;
        $atsv[$pos+1] = $lettermap{$atsv[$pos+1]} if(defined $lettermap{$atsv[$pos+1]});
        $atsv[$pos+1] =~ /^[0-9\.]+$/ or die "Unfinished line $stsv\n";
        push @{$scores[$i]}, $atsv[$pos+1];
        $vals[$i] = [] if not $vals[$i];
        push @{$vals[$i]}, $atsv[$pos];
    }
}

# for(split(/,/, $COMPS)) {
#    my ($base, $sys) = split(/-/);
# }

foreach my $i (1 .. scalar(@{$scores[0]})) {
    my @line = ($i, (map { $scores[$_]->[$i-1] } (0 .. $#scores)), (map { $vals[$_]->[$i-1] } (0 .. $#vals)), $refs[$i-1]);
    print join("\t", @line)."\n";
}
my @avgscores = map { sum(@$_) / $lines } @scores;
print join("\t", "Total", map { sprintf("%.2f", $_) } @avgscores)."\n";
foreach my $lev (2 .. 5) {
    my @levs = map { sprintf("%.2f", sum(map { ($_ >= $lev) ? 1 : 0 } @$_) * 100 / $lines) } @scores;
    print join("\t", "$lev+", @levs)."\n";
}
foreach my $i (0 .. $#avgscores-1) {
    foreach my $j ($i+1 .. $#avgscores) {
        my ($w, $t, $l) = bootstrap($scores[$i], $scores[$j]);
        print "$i-vs.-$j\t$w\t$t\t$l\n";
    }
}

# Perform bootstrap resampling to estimate probabilities
sub bootstrap {
    my $ITER = 10000;
    my @sysa = @{shift(@_)};
    my @sysb = @{shift(@_)};
    @sysa == @sysb or die "Uneven numbers of scores";
    my @ids = (0 .. $#sysa);
    my @wins = (0, 0, 0);
    for(1 .. $ITER) {
        @ids = shuffle(@ids);
        my ($sa, $sb);
        for(@ids[0 .. int(@ids/2)]) {
            $sa += $sysa[$_];
            $sb += $sysb[$_];
            # print "$sysa[$_] $sysb[$_]\n";
        }
        if($sa > $sb) { $wins[0]++; }
        elsif($sb > $sa) { $wins[2]++; }
        else { $wins[1]++; }
    }
    return map { $_ / $ITER } @wins;
}

File: syseval-report.py
import sys

import numpy as np

"""
What's this?
------------

Given a TSV-format manual evaluation file created by syseval-combine.py,
this script restores the orders of the systems and calculates an average
score for each system.

Usage and Data Format
---------------------

syseval-report.py finished-eval.tsv output.ids

Where finished-eval is a completely filled in manual evaluation TSV file
and output.ids is the ID file output by syseval-combine.pl
"""

with open(sys.argv[1], 'r') as tsv_in, open(sys.argv[2], 'r') as ids_in:
    header = next(tsv_in).split('\t')
    if header[0] != 'Source':
        raise ValueError(f'Illegal header {header}')
    has_reference = 1 if header[1] == 'Reference' else 0
    parsed_tsv = []
    curr_data = {}
    for line in tsv_in:
        cols = [x.strip() for x in line.split('\t')]
        if len(cols) > 1:
            if 'src' not in curr_data:
                curr_data['src'] = cols[0]
                if has_reference:
                    curr_data['ref'] = cols[1]
                curr_data['hyps'] = []
                curr_data['scores'] = []
            curr_data['hyps'].append(cols[1+has_reference])
            curr_data['scores'].append(cols[2+has_reference])
        else:
            parsed_tsv.append(curr_data)
            curr_data = {}
    all_scores = None
    for ids_str, data in zip(ids_in, parsed_tsv):
        ids = [int(x) for x in ids_str.strip().split('\t')]
        if not all_scores:
            all_scores = [list() for _ in ids]
        my_hyps = [data['hyps'][i] for i in ids]
        my_scores = [data['scores'][i] for i in ids]
        out_cols = [data['src']]
        if 'ref' in data:
            out_cols.append(data['ref'])
        for i, (hyp, score) in enumerate(zip(my_hyps, my_scores)):
            out_cols += [hyp, score]
            try:
                all_scores[i].append(float(score))
            except:
                print(f'WARNING: illegal score {score}', file=sys.stderr)
        print('\t'.join(out_cols))
    avg_scores = [str(np.average(x)) for x in all_scores]
    print('\n--- Average Scores ---\n'+'\t'.join(avg_scores))

